sklearn/metrics/pairwise.py

```diff
@@ -1182,17 +1182,22 @@ def _parallel_pairwise(X, Y, func, n_jobs, **kwds):
 
     if Y is None:
         Y = X
+    X, Y, dtype = _return_float_dtype(X, Y)
 
     if effective_n_jobs(n_jobs) == 1:
         return func(X, Y, **kwds)
 
     # enforce a threading backend to prevent data communication overhead
     fd = delayed(_dist_wrapper)
-    ret = np.empty((X.shape[0], Y.shape[0]), dtype=X.dtype, order='F')
+    ret = np.empty((X.shape[0], Y.shape[0]), dtype=dtype, order='F')
     Parallel(backend="threading", n_jobs=n_jobs)(
         fd(func, ret, s, X, Y[s], **kwds)
         for s in gen_even_slices(_num_samples(Y), effective_n_jobs(n_jobs)))
 
+    if ((X is Y or Y is None) and func is euclidean_distances):
```

Comment:
why ``func is euclidean_distances``? what about other metrics?

Reply:
when the metric is a callable, we don't know if it's a real metric, i.e. symmetric. It has been decided in #12612 to only fix it for euclidean for now, and maybe enforce real metrics with a deprecation cycle, but that would be for later. For now I just wanted to a consistent results for metrics between `pairwise_distances`, `pairwise_distances_chunked`, all metrics, and all n_jobs.

Reply:
I was a bit confused too. Maybe simply add a comment like "other metrics are more numerically stable"